### ISSUE TITLE
Switching not working when connected to external display

### ISSUE THREAD

**Issue by lewisxy** [Askannz/optimus-manager#343]
**Describe the bug**
Switching works fine when only using internal display. However, when plugged into an external display (via HDMI), switching became extremely slow (took over several minutes, sometimes forever and I have to restart the computer, stuck in loading screen of Manjaro after entering login password), and eventually failed (always fallback to intel mode, which does not support external display). The issue persisted even after unplugged the display (after one switching failure, all the rest of switching will fail).

**System info**
Manjaro (Linux Kernel 5.9.8)
KDE plasma
Dell Precision M4800 (with NVIDIA Quadra K2100M using video-hybrid-intel-nvidia-450xx-prime driver)
optimus-manager (version 1.3 installed with pacman)
default config file (I configure the system mostly following [this](https://archived.forum.manjaro.org/t/guide-install-and-configure-optimus-manager-for-hybrid-gpu-setups-intel-nvidia/92196) guide)

**Logs**
Run `optimus-manager --status` in a console, and if an error message appears, post it here. The message should also point you to a log path, attach it here as well. If there is no error message but you are still experiencing issues, grab the most recent files in `/var/log/optimus-manager/switch/` and `/var/log/optimus-manager/daemon/`.
```
ERROR: the latest GPU setup attempt failed at Xorg pre-start hook.
```
switching log
```
[10] INFO: # Xorg pre-start hook
[10] INFO: Previous state was: {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': 'nvidia'}
[10] INFO: Requested mode is: intel
[10] INFO: Checking for GDM display servers
[229] INFO: Available modules: ['nouveau', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[229] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
[330] ERROR: Xorg pre-start setup error
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/bash.py", line 11, in exec_bash
    out = subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['bash', '-c', 'modprobe -r nvidia_drm nvidia_modeset nvidia_uvm nvidia']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 196, in _unload_modules
    exec_bash("modprobe -r " + " ".join(modules_to_unload))
  File "/usr/lib/python3.8/site-packages/optimus_manager/bash.py", line 18, in exec_bash
    raise BashError(
optimus_manager.bash.BashError: Failed to execute 'modprobe -r nvidia_drm nvidia_modeset nvidia_uvm nvidia' :
modprobe: FATAL: Module nvidia is in use.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/hooks/pre_xorg_start.py", line 45, in main
    setup_kernel_state(config, prev_state, requested_mode)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 24, in setup_kernel_state
    _nvidia_down(config)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 65, in _nvidia_down
    _unload_nvidia_modules(available_modules)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 148, in _unload_nvidia_modules
    _unload_modules(available_modules, ["nvidia_drm", "nvidia_modeset", "nvidia_uvm", "nvidia"])
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 198, in _unload_modules
    raise KernelSetupError("Cannot unload modules %s : %s" % (str(modules_to_unload), str(e)))
optimus_manager.kernel.KernelSetupError: Cannot unload modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] : Failed to execute 'modprobe -r nvidia_drm nvidia_modeset nvidia_uvm nvidia' :
modprobe: FATAL: Module nvidia is in use.

[332] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
[332] INFO: Writing state {'type': 'pre_xorg_start_failed', 'switch_id': '20201119T020703', 'requested_mode': 'intel'}
```

**Comment by WhyDoWeWonder** [comment]
Have you checked to see if any process is using the nvidia module before you switch? Try running `lsof /dev/nvidia*` and seeing if that outputs anything.

**Comment by lewisxy** [comment]
When connected to external display. Here is the output of that command.
```
COMMAND    PID    USER   FD   TYPE  DEVICE SIZE/OFF NODE NAME
plasmashe 8461 lewisxy  mem    CHR 195,255           619 /dev/nvidiactl
plasmashe 8461 lewisxy  mem    CHR   195,0           620 /dev/nvidia0
plasmashe 8461 lewisxy   15u   CHR 195,255      0t0  619 /dev/nvidiactl
plasmashe 8461 lewisxy   16u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 8461 lewisxy   17u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   18u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   19u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 8461 lewisxy   21u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   22u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   23u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   24u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   25u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   33u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 8461 lewisxy   34u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   35u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   36u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   37u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   39u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 8461 lewisxy   40u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   41u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   42u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   43u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   44u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   61u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 8461 lewisxy   62u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   63u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   64u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   65u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   66u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 8461 lewisxy   71u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8689 lewisxy  mem    CHR 195,255           619 /dev/nvidiactl
thumbnail 8689 lewisxy  mem    CHR   195,0           620 /dev/nvidia0
thumbnail 8689 lewisxy    8u   CHR 195,255      0t0  619 /dev/nvidiactl
thumbnail 8689 lewisxy    9u   CHR 195,254      0t0  621 /dev/nvidia-modeset
thumbnail 8689 lewisxy   10u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8689 lewisxy   11u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8689 lewisxy   12u   CHR 195,254      0t0  621 /dev/nvidia-modeset
thumbnail 8689 lewisxy   14u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8689 lewisxy   15u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8689 lewisxy   16u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8689 lewisxy   17u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8689 lewisxy   18u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8720 lewisxy  mem    CHR 195,255           619 /dev/nvidiactl
thumbnail 8720 lewisxy  mem    CHR   195,0           620 /dev/nvidia0
thumbnail 8720 lewisxy    8u   CHR 195,255      0t0  619 /dev/nvidiactl
thumbnail 8720 lewisxy    9u   CHR 195,254      0t0  621 /dev/nvidia-modeset
thumbnail 8720 lewisxy   10u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8720 lewisxy   11u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8720 lewisxy   12u   CHR 195,254      0t0  621 /dev/nvidia-modeset
thumbnail 8720 lewisxy   14u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8720 lewisxy   15u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8720 lewisxy   16u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8720 lewisxy   17u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 8720 lewisxy   18u   CHR   195,0      0t0  620 /dev/nvidia0
```
I then unplugged the external display and switch to Intel, the above command output nothing.
I then switched back to Nvidia (without plugging into external display), and here is the output.
```
COMMAND     PID    USER   FD   TYPE  DEVICE SIZE/OFF NODE NAME
plasmashe 24430 lewisxy  mem    CHR 195,255           619 /dev/nvidiactl
plasmashe 24430 lewisxy  mem    CHR   195,0           620 /dev/nvidia0
plasmashe 24430 lewisxy   15u   CHR 195,255      0t0  619 /dev/nvidiactl
plasmashe 24430 lewisxy   16u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 24430 lewisxy   17u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   18u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   19u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 24430 lewisxy   21u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   22u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   23u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   24u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   25u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   29u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 24430 lewisxy   30u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   31u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   32u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   33u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   34u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   58u   CHR 195,254      0t0  621 /dev/nvidia-modeset
plasmashe 24430 lewisxy   59u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   60u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   61u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   62u   CHR   195,0      0t0  620 /dev/nvidia0
plasmashe 24430 lewisxy   63u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24671 lewisxy  mem    CHR 195,255           619 /dev/nvidiactl
thumbnail 24671 lewisxy  mem    CHR   195,0           620 /dev/nvidia0
thumbnail 24671 lewisxy    8u   CHR 195,255      0t0  619 /dev/nvidiactl
thumbnail 24671 lewisxy    9u   CHR 195,254      0t0  621 /dev/nvidia-modeset
thumbnail 24671 lewisxy   10u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24671 lewisxy   11u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24671 lewisxy   12u   CHR 195,254      0t0  621 /dev/nvidia-modeset
thumbnail 24671 lewisxy   14u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24671 lewisxy   15u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24671 lewisxy   16u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24671 lewisxy   17u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24671 lewisxy   18u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24702 lewisxy  mem    CHR 195,255           619 /dev/nvidiactl
thumbnail 24702 lewisxy  mem    CHR   195,0           620 /dev/nvidia0
thumbnail 24702 lewisxy    8u   CHR 195,255      0t0  619 /dev/nvidiactl
thumbnail 24702 lewisxy    9u   CHR 195,254      0t0  621 /dev/nvidia-modeset
thumbnail 24702 lewisxy   10u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24702 lewisxy   11u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24702 lewisxy   12u   CHR 195,254      0t0  621 /dev/nvidia-modeset
thumbnail 24702 lewisxy   14u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24702 lewisxy   15u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24702 lewisxy   16u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24702 lewisxy   17u   CHR   195,0      0t0  620 /dev/nvidia0
thumbnail 24702 lewisxy   18u   CHR   195,0      0t0  620 /dev/nvidia0
```
I am not exactly sure what those processes are as they keep changing, but one of them is the `/usr/bin/plasmashell`.

**Comment by WhyDoWeWonder** [comment]
Doesn't seem to be it, I have a very similar list of processes, are there any applications you can think of that might be interacting with the nvidia driver? For me it was thinkfan which was reading the temperature value from the proprietary driver.

**Comment by michaldybczak** [comment]
I believe I have a similar problem, because the symptoms are identical, but first I have to solve the problem with the black screens on hybrid mode with secondary monitor, then go back to this one.

In the past this happened once in a while, so it could be ignored, but after recent updates every switch on multi-monitor seems to be stuck and never succeeds, so I have to reboot the system. No logs come out of it because all seems fine on their side, so I assume this switch isn't even written. The problem escalated.

Once I solve the issue with multi-monitors working in hybrid setup (with 455 drivers), I may do some testing with time stamps to evaluate logs.

**Comment by blacktoz** [comment]
> I believe I have a similar problem, because the symptoms are identical, but first I have to solve the problem with the black screens on hybrid mode with secondary monitor, then go back to this one.
> 
> In the past this happened once in a while, so it could be ignored, but after recent updates every switch on multi-monitor seems to be stuck and never succeeds, so I have to reboot the system. No logs come out of it because all seems fine on their side, so I assume this switch isn't even written. The problem escalated.
> 
> Once I solve the issue with multi-monitors working in hybrid setup (with 455 drivers), I may do some testing with time stamps to evaluate logs.

does it work now ?

**Comment by nwildner** [comment]
@lewisxy @michaldybczak are you stil facing this issue?

Since the last update on this case was from 3 years ago, we'll keep this case opened for some days and close it if no answer is provided.

Have a nice week ahead.

**Comment by michaldybczak** [comment]
Since the laptop with Nvidia is the one I rarely use nowadays (I moved on to one with AMD dGPU), I had to update it first, then install optimus-manager-git.

Old optimus-manager gave me no issue since a long time and secondary monitor was working.

After the update and reboot, Plasma asked me what to do with the second screen, I chose to extend it to the left and... nothing happened. The secondary screen showed "No HDMI signal". This behaves identically to a pure hybrid setup on Manjaro where the secondary monitor doesn't work, regardless of type of a session (X11 or Wayland).

I spend in the past a lot of time trying to figure out the issue but without results. The optimus-manager was the only way to get the secondary monitor working on hybrid session in Wayland (although it supposes not influence Wayland session, but it did work there as well...) and even in X11.

I tried to use the same config for Nvidia that optimus-manager was creating - no change. I added all packages that were installed with optimus-manager - still nothing. Tried various configs, red Arch wiki about Wayland and monitors, but it lead me nowhere. I asked people around, but nobody could help me. So I had to go back to optimus-manager and its unknown magic that was making my secondary screen work.

Now, it seems that the optimus-manager-git package from AUR gives me the same problem - hybrid mode behaves as hybrid setup (default for hybrid laptops with Nvidia on Manjaro). However, in the past, I could get my secondary screen by going back to the optimus-manager, but this time, when I did it, it didn't fix the issue. So I'm restoring my backup and will stay on old optimus-manager.

I don't think my issues have anything to do with the issue here.

Since old optimus-manager works for me, I will probably put it on pacman ignore list and will see how long it will hold working that way. At the moment, it's the only way for me. I wished I knew what it was doing to make my screen work, but I don't and despite long search, I didn't find any solution.
